lib/cost_report.rb

```diff
@@ -1,15 +1,89 @@
 # frozen_string_literal: true
 
 require "ht_item_overlap"
+require "utils/waypoint"
+require "services"
 
 # Generates reports based on h_share
 class CostReport
-  attr_accessor :organization, :freq_table
+  attr_accessor :organization, :logger, :maxlines, :target_cost
 
-  def initialize(org = nil)
+  def initialize(org = nil, cost: ENV["target_cost"], lines: 1_000_000, logger: Services.logger)
     @organization = org
-    # { org => { 1 org : count, 2 org : count }
-    @freq_table = Hash.new {|hash, key| hash[key] = Hash.new(0) }
+    @target_cost = cost.to_f
+    @maxlines = lines
+    @logger = logger
+    # Member Hash of a format hash of a member count hash
+    # { org => { ser : { 1 org : count, 2 org : count }, mpm : {...
+    @freq_table = Hash.new do |hash, member|
+      hash[member] = Hash.new {|fmt_hash, fmt| fmt_hash[fmt] = Hash.new(0) }
+    end
+  end
+
+  def num_volumes
+    @num_volumes ||= Cluster.collection.aggregate(
+      [
+        { '$match': { "ht_items.0": { "$exists": 1 } } },
+        { '$group': { _id: nil, items_count: { "$sum": { '$size': "$ht_items" } } } }
+      ]
+    ).first[:items_count]
+  end
+
+  def num_pd_volumes
+    @num_pd_volumes ||= Cluster.collection.aggregate(
+      [
+        { '$match': { "ht_items.0": { "$exists": 1 } } },
+        { '$group': { _id:         nil,
+                      items_count: { "$sum": { "$size": {
+                        "$filter": { input: "$ht_items",
+                                     as:    "item",
+                                     cond:  { "$eq": ["$$item.access", "allow"] } }
+                      } } } } }
+      ]
+    ).first[:items_count]
+  end
+
+  def cost_per_volume
+    target_cost / num_volumes.to_f
+  end
+
+  def total_weight
+    Services.ht_members.members.map {|_id, member| member.weight }.sum
+  end
+
+  def pd_cost
+    cost_per_volume * num_pd_volumes
+  end
+
+  def pd_cost_for_member(member)
+    member_weight     = Services.ht_members[member.to_s].weight
+    (pd_cost / total_weight) * member_weight
+  end
+
+  def freq_table
+    compile_frequency_table unless @freq_table.any?
+    @freq_table
+  end
+
+  def compile_frequency_table
+    waypoint = Utils::Waypoint.new(maxlines)
+    logger.info("Begin compiling hscore frequency table.")
+    matching_clusters.each do |c|
+      c.ht_items.each do |ht_item|
+        next unless ht_item.access == "deny"
+
+        waypoint.incr
+        add_ht_item_to_freq_table(ht_item)
+        waypoint.on_batch {|wp| logger.info wp.batch_line }
+      end
+    end
+  end
+
+  def add_ht_item_to_freq_table(ht_item)
+    overlap = HtItemOverlap.new(ht_item)
+    overlap.matching_orgs.each do |org|
+      @freq_table[org.to_sym][ht_item._parent.format.to_sym][overlap.matching_orgs.count] += 1
```

Comment:
I would move the `#to_sym` calls into the definition of `freq_table`; easier to not forget them.

Reply:
freq_table is a stupid doubly nested hash. 

I tried turning this:
` @freq_table = Hash.new do |hash, member|
  hash[member] = Hash.new {|fmt_hash, fmt| fmt_hash[fmt] = Hash.new(0) }
end`

into this:
`@freq_table = Hash.new do |hash, member|
    hash[member.to_sym] = Hash.new {|fmt_hash, fmt| fmt_hash[fmt.to_sym] = Hash.new(0) }
end`

which doesn't work. The alternative would be a smarter `class FreqTable` which may or may not be worth doing.

Reply:
Are you sure? Not that it's a huge deal either way...

```ruby
f = Hash.new do |hash, member|
  hash[member.to_sym] = Hash.new do |fmt_hash, fmt|
    fmt_hash[fmt.to_sym] = Hash.new(0)
  end
end
f["bill"]["book"]["umich"] = 2

ap f
{
    :bill => {
        :book => {
            "umich" => 2
        }
    }
}
```

Reply:
If you try incrementing f["bill"] it won't find it, creates a new hash, and assigns it to f["bill".to_sym]. I think. Anyway, the specs failed when I tried it.